File: src-tauri/src/platform.rs
use serde::Serialize;

#[derive(Debug, Clone, Copy, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct PlatformConfig {
    pub id: &'static str,
    pub label: &'static str,
    pub emulator_name: &'static str,
    pub executable_hint: &'static str,
    pub launch_args_template: &'static str,
    pub system_path_hint: &'static str,
    pub expected_extensions: &'static [&'static str],
}

pub const MVP_PLATFORM_IDS: &[&str] = &["switch", "ps1", "ps2", "gba", "nes"];

pub const MVP_PLATFORM_CONFIGS: &[PlatformConfig] = &[
    PlatformConfig {
        id: "switch",
        label: "Nintendo Switch",
        emulator_name: "Ryujinx",
        executable_hint: "Ryujinx.exe",
        launch_args_template: "{game_path}",
        system_path_hint: "portable/system",
        expected_extensions: &[".nsp", ".xci"],
    },
    PlatformConfig {
        id: "ps1",
        label: "PlayStation 1",
        emulator_name: "DuckStation",
        executable_hint: "duckstation-qt-x64-ReleaseLTCG.exe",
        launch_args_template: "-batch \"{game_path}\"",
        system_path_hint: "bios",
        expected_extensions: &[".bin", ".cue", ".iso", ".img", ".pbp"],
    },
    PlatformConfig {
        id: "ps2",
        label: "PlayStation 2",
        emulator_name: "PCSX2",
        executable_hint: "pcsx2-qt.exe",
        launch_args_template: "-fullscreen -- \"{game_path}\"",
        system_path_hint: "bios",
        expected_extensions: &[".iso", ".bin", ".img"],
    },
    PlatformConfig {
        id: "gba",
        label: "Game Boy Advance",
        emulator_name: "mGBA",
        executable_hint: "mGBA.exe",
        launch_args_template: "{game_path}",
        system_path_hint: "",
        expected_extensions: &[".gba"],
    },
    PlatformConfig {
        id: "nes",
        label: "NES / Famicom",
        emulator_name: "Mesen",
        executable_hint: "Mesen.exe",
        launch_args_template: "{game_path}",
        system_path_hint: "",
        expected_extensions: &[".nes"],
    },
];

pub fn platform_config(platform: &str) -> Option<&'static PlatformConfig> {
    MVP_PLATFORM_CONFIGS
        .iter()
        .find(|config| config.id == platform)
}

pub fn is_mvp_platform(platform: &str) -> bool {
    MVP_PLATFORM_IDS.contains(&platform)
}

pub fn default_launch_args_template(platform: &str) -> Option<&'static str> {
    platform_config(platform).map(|config| config.launch_args_template)
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn mvp_platform_matrix_has_expected_defaults() {
        assert!(is_mvp_platform("switch"));
        assert!(!is_mvp_platform("dreamcast"));
        assert_eq!(default_launch_args_template("ps1"), Some("-batch \"{game_path}\""));
        assert_eq!(platform_config("ps2").map(|config| config.executable_hint), Some("pcsx2-qt.exe"));
    }
}

File: src/types/platform.ts
export const PLATFORMS = [
  'switch',
  'nes',
  'snes',
  'n64',
  'gba',
  'nds',
  'gamecube',
  'wii',
  'ps1',
  'ps2',
  'ps3',
  'psp',
  'genesis',
  'saturn',
  'dreamcast'
] as const;

export type Platform = (typeof PLATFORMS)[number];

export const MVP_PLATFORMS = ['switch', 'ps1', 'ps2', 'gba', 'nes'] as const satisfies readonly Platform[];
export type MvpPlatform = (typeof MVP_PLATFORMS)[number];

export const EMULATOR_MANAGER_PLATFORMS = ['nes', 'snes', 'n64', 'gba', 'ps2', 'psp', 'ps1', 'switch'] as const satisfies readonly Platform[];
export type EmulatorManagerPlatform = (typeof EMULATOR_MANAGER_PLATFORMS)[number];

export const PLATFORM_LABELS: Record<Platform, string> = {
  switch: 'Nintendo Switch',
  nes: 'NES / Famicom',
  snes: 'Super Nintendo',
  n64: 'Nintendo 64',
  gba: 'Game Boy Advance',
  nds: 'Nintendo DS',
  gamecube: 'GameCube',
  wii: 'Wii',
  ps1: 'PlayStation 1',
  ps2: 'PlayStation 2',
  ps3: 'PlayStation 3',
  psp: 'PSP',
  genesis: 'Sega Genesis / Mega Drive',
  saturn: 'Sega Saturn',
  dreamcast: 'Sega Dreamcast'
};

export const PLATFORM_DEFAULT_EXTENSIONS: Record<Platform, string[]> = {
  switch: ['.nsp', '.xci', '.nca'],
  nes: ['.nes'],
  snes: ['.sfc', '.smc'],
  n64: ['.z64', '.n64', '.v64'],
  gba: ['.gba'],
  nds: ['.nds'],
  gamecube: ['.rvz', '.iso', '.gcm'],
  wii: ['.rvz', '.wbfs', '.iso'],
  ps1: ['.cue', '.bin', '.iso', '.img', '.pbp'],
  ps2: ['.iso', '.bin', '.img'],
  ps3: ['.pkg', '.iso'],
  psp: ['.iso', '.cso', '.pbp'],
  genesis: ['.md', '.bin', '.gen', '.smd'],
  saturn: ['.cue', '.bin', '.iso', '.chd'],
  dreamcast: ['.gdi', '.cdi', '.chd', '.iso']
};

export const PLATFORM_EMULATOR_HINTS: Record<MvpPlatform, string> = {
  switch: 'eden.exe',
  ps1: 'duckstation-qt-x64-ReleaseLTCG.exe',
  ps2: 'pcsx2-qt.exe',
  gba: 'mGBA.exe',
  nes: 'Mesen.exe'
};

export const PLATFORM_DEFAULT_LAUNCH_ARGS: Record<MvpPlatform, string> = {
  switch: '{game_path}',
  ps1: '-batch "{game_path}"',
  ps2: '-fullscreen -- "{game_path}"',
  gba: '-f {game_path}',
  nes: '{game_path}'
};

export function isMvpPlatform(platform: string): platform is MvpPlatform {
  return (MVP_PLATFORMS as readonly string[]).includes(platform);
}
